src/NuGet.Core/NuGet.Packaging/Signing/Verification/SignatureTrustAndValidityVerificationProvider.cs

```diff
@@ -26,45 +25,109 @@ public SignatureTrustAndValidityVerificationProvider()
         {
             token.ThrowIfCancellationRequested();
-
-            var result = VerifyValidityAndTrust(signature, settings);
+            var result = VerifySignatureAndCountersignature(signature, settings);
             return Task.FromResult(result);
         }
 
 #if IS_DESKTOP
-        private PackageVerificationResult VerifyValidityAndTrust(PrimarySignature signature, SignedPackageVerifierSettings settings)
+        private PackageVerificationResult VerifySignatureAndCountersignature(
+            PrimarySignature signature,
+            SignedPackageVerifierSettings settings)
+        {
+            var issues = new List<SignatureLog>();
+            var certificateExtraStore = signature.SignedCms.Certificates;
+            var primarySignatureHasCountersignature = SignatureUtility.HasRepositoryCountersignature(signature);
+
+            // Only accept untrusted root if the signature has a countersignature that we can validate against
+            var verifySettings = new SignatureVerifySettings(
+                treatIssuesAsErrors: !settings.AllowIllegal,
+                allowUntrustedRoot: primarySignatureHasCountersignature,
+                allowUnknownRevocation: settings.AllowUnknownRevocation,
+                logOnSignatureExpired: !primarySignatureHasCountersignature);
+
+            var primarySignatureVerificationSummary = VerifyValidityAndTrust(signature, settings, verifySettings, certificateExtraStore, issues);
+            var status = primarySignatureVerificationSummary.Status;
+
+            if (primarySignatureHasCountersignature)
+            {
+                if (settings.AlwaysVerifyCountersignature || ShouldFallbackToRepositoryCountersignature(primarySignatureVerificationSummary))
+                {
+                    var counterSignature = RepositoryCountersignature.GetRepositoryCountersignature(signature);
+                    verifySettings = new SignatureVerifySettings(
+                        treatIssuesAsErrors: !settings.AllowIllegal,
+                        allowUntrustedRoot: false,
+                        allowUnknownRevocation: settings.AllowUnknownRevocation,
+                        logOnSignatureExpired: true);
+
+                    var countersignatureVerificationSummary = VerifyValidityAndTrust(counterSignature, settings, verifySettings, certificateExtraStore, issues);
+                    status = countersignatureVerificationSummary.Status;
+
+                    if (!Rfc3161TimestampVerificationUtility.ValidateSignerCertificateAgainstTimestamp(signature.SignerInfo.Certificate, countersignatureVerificationSummary.Timestamp))
+                    {
+                        issues.Add(SignatureLog.Issue(!settings.AllowIllegal, NuGetLogCode.NU3011, string.Format(CultureInfo.CurrentCulture, Strings.SignatureNotTimeValid, nameof(AuthorPrimarySignature))));
```

Comment:
`nameof(AuthorPrimarySignature)` is not generally correct.  `nameof(PrimarySignature)` would be more generally correct; however, consider my previous advice before changing this.